src/main/scala/com/fulcrumgenomics/basecalling/ExtractBasecallingParams.scala

```diff
@@ -0,0 +1,191 @@
+    |with name "barcode_params.lane-<lane>.txt".
+    |
+    |The library parameter file is used by Picard's IlluminaBasecallsToSam to demultiplex samples and name each sample's
+    |output BAM file.  The parameter file will be written to the output directory with name
+    |"library_params.lane-<lane>.txt".  The path to each sample's BAM file will be specified in the library parameter
+    |file.  Each BAM file will have path "<output>/<sample-name>.<library-id>.bam" if the library identifier is given or
+    |"<output>/<sample-name>.bam" if library identifer is not specified in the sample sheet.
+  """)
+class ExtractBasecallingParams
+(
+  @arg(flag="i", doc="The sample sheet.") val sampleSheet: FilePath,
+  @arg(flag="o", doc="The output folder to where lane-specific data should be written.") val output: DirPath,
+  @arg(flag="l", doc="The lane (1-based).") val lane: Int
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(sampleSheet)
+  Io.assertWritableDirectory(output)
+  validate(0 < lane, "--lane must be greater than zero")
+
+  override def execute(): Unit = {
+    val params = BasecallingParams.from(sampleSheet=SampleSheet(sampleSheet=sampleSheet), lane=lane, output=output)
+    logger.info(s"Barcode params written to: ${params.barcodeFile}")
+    logger.info(s"Library params written to: ${params.libraryParamsFile}")
+  }
+}
+
+
+/** A small class to store where the basecalling parameter and BAM files are located. */
+case class BasecallingParams(barcodeFile: FilePath, libraryParamsFile: FilePath, bams: Seq[PathToBam])
+
+/** Contains methods useful for extractin basecalling paramter information. */
+object BasecallingParams {
+
+  /** Extracts sample and library information from an sample sheet for a given lane.  The resulting files are used
+    * to run Picard's ExtractIlluminaBarcodes, CollectIlluminaBasecallingMetrics, and IlluminaBasecallsToSam. See
+    * [[ExtractBasecallingParams]] for more information.
+    *
+    * @param sampleSheet the Illumina Experiment Manager sample sheet.
+    * @param lane the lane number.
+    * @param output the output directory where the barcode, library param, and BAM files should be written.
+    */
+  def from(sampleSheet: SampleSheet, lane: Int, output: DirPath): BasecallingParams = {
+    require(0 < lane)
+
+    val barcodeLines      = new ListBuffer[String]()
+    val libraryParamLines = new ListBuffer[String]()
+    val samples           = sampleSheet.filter(_.lane.forall(_ == lane)).toSeq
+    val barcodeFile       = barcodeFileFrom(output=output, lane=lane)
+    val libraryParamsFile = libraryParamsFileFrom(output=output, lane=lane)
+
+    // Dual vs single indexed
+    val dualIndexed = samples.exists(_.i5IndexBases.nonEmpty)
+
+    // Check for project/description on the sample
+    val includeDescription    = samples.exists(s => s.project.nonEmpty || s.description.nonEmpty)
```

Comment:
I don't love this.  I've seen a lot of labs re-use the "project" field for things that have no place in a BAM file (instead of creating their own fields).  E.g. one lab puts the email address of the person who made the libraries in that field!!